bundles/org.openhab.binding.growatt/src/main/java/org/openhab/binding/growatt/internal/action/GrowattActions.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.growatt.internal.action;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.growatt.internal.handler.GrowattInverterHandler;
+import org.openhab.core.automation.annotation.ActionInput;
+import org.openhab.core.automation.annotation.RuleAction;
+import org.openhab.core.thing.binding.ThingActions;
+import org.openhab.core.thing.binding.ThingActionsScope;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the {@link ThingActions} interface used for setting up battery charging and discharging programs.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@ThingActionsScope(name = "growatt")
+@NonNullByDefault
+public class GrowattActions implements ThingActions {
+
+    private final Logger logger = LoggerFactory.getLogger(GrowattActions.class);
+    private @Nullable GrowattInverterHandler handler;
+
+    public static void setupChargingProgram(ThingActions actions, Number chargingPower, Number targetSOC,
+            boolean allowAcCharging, String startTime, String stopTime, boolean programEnable) {
+        ((GrowattActions) actions).setupChargingProgram(chargingPower, targetSOC, allowAcCharging, startTime, stopTime,
+                programEnable);
+    }
+
+    public static void setupDischargingProgram(ThingActions actions, Number dischargingPower, Number targetSOC,
+            String startTime, String stopTime, boolean programEnable) {
+        ((GrowattActions) actions).setupDischargingProgram(dischargingPower, targetSOC, startTime, stopTime,
```

Comment:
Is this unckecked cast safe? Why is the parameter type ThingActions and not GrowattActions?

Reply:
See above